### DISCUSSION TITLE
The right approach to use slang as a formatter

### DISCUSSION THREAD

**Discussion by suzizecat** [MikePopoloski/slang#828] [Q&A] [answered]
Hi there !

I am currently starting to work on a SystemVerilog source formatter tool (alongside a language server) using slang.
For the time being, the idea would be to format the code to my liking (before adding any options). 
So, to keep it simple:
 - Perform indentation,
 - Beautify spacing (remove multiple spaces, aligns statements, that sort of stuff). 

I am also allowing myself to move around bits of code to make it easier on my eyes (one statement per line, in example) and easier to implement.

My approach is to use a custom visitor (class derived from `slang::syntax::SyntaxVisitor` ), reads the AST and actually rewrite the source code from this. 

My problem is that it seems quite complicated to handle comments, empty lines and anything that is not directly represented by a token. (the trivia, I guess ?) 
Moreover, I find it particularly tedious to handle all types of possible syntax node on each and every situation.
In example, for `parameters` declaration, the type of the parameter is a `DataTypeSyntax` which can be 8 different actual objects with various structures. I am aware of the `SyntaxNode::toString()` function, but it pose some challenges for formatting (newlines and leading spaces are kept).

While I am not surprised by the tediousness of the job, I am wondering if I am going down the right path or if someone did such thing with a more clever approach that I am not aware of. 

Best regards !

**Comment by MikePopoloski** [comment]
The SyntaxRewriter class (which inherits from SyntaxVisitor) might be helpful here, but I agree that it's probably going to be some work to make it less tedious. Note that you don't necessarily have to write a handler for every type of syntax node; you can write handlers for the base classes (like DataTypeSyntax) and that will be called for all types of nodes that don't have more specific handlers.

If I were starting this project I would look at how clang-format works, and also probably how some of the C# auto-formatter tools work (slang's parser design is inspired by clang and the Roslyn C# compiler). If you want to add helpers / utilities / make upstream changes to make the job of writing the formatter easier I'm happy to take PRs.

**Reply by suzizecat** [reply]
> Note that you don't necessarily have to write a handler for every type of syntax node; you can write handlers for the base classes (like DataTypeSyntax) and that will be called for all types of nodes that don't have more specific handlers.

I meant that rather in the formatting part rather than for the visitor. Like, I have to handle specifically if a parameter value is an integer, a numeric expression (`3*4`) a litteral string or whatever.

> If I were starting this project I would look at how clang-format works, and also probably how some of the C# auto-formatter tools work (slang's parser design is inspired by clang and the Roslyn C# compiler)

Thanks for the pointers, I'll have a look. 

By any chance, is there an easy way to have the raw list of typed tokens of a file (I believe just after lexing) ? Meaning a `vector` (or equivalent) of only Tokens annotated with their types.  Ideally with whitespaces, newlines, comments and such still in a token form instead of trivia (to keep the order between such tokens and the "meaningful" / actual RTL tokens) 
I am not sure how I will actually work so I am exploring a bit 🙂

**Reply by MikePopoloski** [reply]
Well you can certainly run just the lexer and get a list of tokens, but you won't know the larger language structures those tokens are a part of until you run the parser. Trivia (whitespace, comments, etc) are never in token form; they're always trivia. Tokens are only things that have actual impact on the parsing / semantics of the design.

**Reply by suzizecat** [reply]
Ok, thanks for the clear answers 🙂 
Just to be sure, could you confirm that trivia are (always):
 - Preceding the token on which they are "attached"
 - Always stored "in order" (first trivia encountered is the first in the trivia list) 

So that, in order to print the code as it was in the original file you could do something such as:
```cpp
for (const Trivia& triv : token.trivia)
{
    cout << triv; // Trivia includes newlines
}
cout << token.rawText();
```
(This is simplified and all from memory since I don't have the actual slang documentation with me right now)

**Reply by MikePopoloski** [reply]
Yes, that's right. That's basically what the SyntaxPrinter class does.

**Comment by suzizecat** [comment]
Hello again,

For reference, after fiddling around quite a bit, I found the following code structure to perform fairly well for my goal (Here, attaching an end of line commend to a port declaration) in a module declaration such as 
```systemverilog
module instruction_counter #(
	parameter K_ADWIDTH = 855, //! Address width
	parameter int K_DWIDTH_IS_VERY_VERY_LONG = 8 + 35, K_TOTO = 5
) (
	input  logic                 i_clk     , //! Input clock
	input  logic                 i_rstn    , //! Reset, active low
	input  logic                 i_incr_en , //! Increment enable
	output logic [K_ADWIDTH-1:0] o_addr    , //! Output address
	input  logic                 i_set_addr, //! Override counter value
	input  logic                 i_jump    , //! Jump one address
	input  logic [K_ADWIDTH-1:0] i_addr      //! Address to set the counter
);
```
I came up with something along the lines of 
```cpp
void VisitorModuleBlackBox::handle(const ImplicitAnsiPortSyntax& port)
{
	// bb.ports hold the list of ports of the module I'm analyzing
	if(bb.ports.size() > 0)
	{
		for(const slang::parsing::Trivia t : port.header->getFirstToken().trivia())
		{
			// Assuming that I will only have one comment, could be refined.
			// in particular by checking the line number with the port declaration.
			if(t.getRawText().starts_with("//"))
			{
				bb.ports.back().comment = t.getRawText();
				break;
			}
		}
	}
	// Do quite a bit of other things...
}
```

I'm starting to get around mixing the visitor strategy (with handles) and manually descending in interesting nodes manually.
It is still tedious though.

**Comment by suzizecat** [comment]
So, after tackling several point on another project of mine, I came back to have a try at this whole formatter thing. 

The issue with my previous approach is the risk of losing elements during the conversion.
In order to avoid this, I am looking into the syntax rewriter class.

My idea would have been to just edit the nodes/tokens/trivia required to perform the wanted changes (In example, add a whitespace trivia before a specified token to perform indentation). 
The issue is that it looks like it is mandatory to rebuild a `SyntaxNode` from scratch in order to be able to perform those edits, which will cause the same risk of losing information if I do not handle a certain type of node.

Would there be a possibility to: 
 1. Perform a deep-copy of a node in a non-const way
 2. Edit this node or its children in whatever way I see fit (mainly changing tokens/trivia) 
 3. Re-inject this new node through the `SyntaxRewriter::replace` utility.

This way, I could be sure that no information is lost without my knowledge.

**Reply by MikePopoloski** [reply]
There is a clone() and a deepClone() method that operates on SyntaxNodes and give you back something mutable that you can change however you want. Does that do what you need?

**Reply by suzizecat** [reply]
That would be exactly what I need !

For reference, this method does not shows up in the documentation (nor in Github search, for that matter) but is indeed there. I will have a try at it !

**Reply by MikePopoloski** [reply]
Ah yeah, it's missing doc comments and nothing shows up in the API reference if it doesn't have doc comments. I'll fix that.

**Reply by suzizecat** [reply]
After looking around, I am starting to get a visitor that extract and present enough data to start actually formatting. 

The issue I have now is that I don't quite see how to create/update trivia. 

The point would be to inject a trivia with whitespaces for indentation/alignments of several parts of the code. However, it looks like I can only create a `Trivia` from a `string_view` and it seems to me that this is not the best course of action as it will be impractical to store a buffer for the `string_view` to refers to. 

Setting aside creating the trivia, could you confirm that the way to update a token in a cloned syntax node with some new trivia would be to use something along the lines of 
```cpp
Token new_token = previous_token.withTrivia(a_bump_allocator,{my_trivia_1, my_trivia_2});
the_syntax_node.anAttribute = new_token;
```
given that the `new_token` will *only* have `my_trivia_1` and `my_trivia_2`.

Also, about the `BumpAllocator`, what should be the lifetime of this object ? 
 1. Should I keep it along until I re-print the (formatted) syntax tree ? 
 2. Can I safely declare it in a local function ? 
 3. Should I keep it until the end of times ? 

As far as I understand, that would be more the first option, but can I safely use it while mixing any kind of things to allocate ? (Is it a kind of "fast malloc" ?) 

-----
PS: 
As of today (manipulating strings through the `fmtlib` for ease of prototyping), my proof of concept performs the following transformation (I have not quite managed the alignment within the ranges):
```systemverilog
// Before
logic simple;
logic [1:0] single_packed;
logic [1:0][3:0] multi_packed;
logic [1:0][K_DIIM-1:0] multi_packed_param;
logic single_unpacked [3:0];

logic  multi_unpacked [3:0][31:0];
logic multi_unpacked_single [3][31];
logic multi_unpacked_weird [3][K_DIMENSION];

logic [1:0][15:0] full_mix [3][31];

//After 
logic                     simple             ;
logic   [1:0]             single_packed      ;
logic   [1:0][3:0]        multi_packed       ;
logic   [1:0][K_DIIM-1:0] multi_packed_param ;
logic                     single_unpacked     [3:0];

logic    multi_unpacked         [3:0][31:0];
logic    multi_unpacked_single  [3][31];
logic    multi_unpacked_weird   [3][K_DIMENSION];

logic   [1:0][15:0] full_mix  [3][31];
```

**Reply by MikePopoloski** [reply]
Yes, that looks like the right way to update the token's trivia.

You need to keep the BumpAllocator alive until anything that could possibly be referring to its memory is gone. Typically this means you no longer have any references to any tokens, syntax nodes, etc. Yes, it's kind of like a fast malloc, and it's even faster to free because everything stays alive together until the allocator is destroyed, and then everything is freed in one go.

Trivia objects are lightweight; the memory for their strings needs to live elsewhere. You can put that memory into the BumpAllocator as well.

**Reply by suzizecat** [reply]
So ! Now the issue is how to declare properly elements using the bump allocator. 

I managed to have a functional PoC but it looks *very* convoluted to me.
Given that `node` is a `DataDeclarationSyntax` that I am visiting, I came up with:
```cpp
slang::syntax::DataDeclarationSyntax* work = deepClone(*node,_mem);
std::cout << "Modifier pre-update: " << work->modifiers.toString() << std::endl;

//Temp token list for modifiers;
std::vector<slang::parsing::Token> tok_list;

//work->modifier is a TokenList
for(const auto& tok : work->modifiers)
{
	std::string spacing{" !-"}; // Will be an actual spacing.

	// Store the string in "persistent" memory (BumpAllocator)
	slang::byte* spacing_base = _mem.allocate(spacing.length(), alignof(char));
	std::memcpy(spacing_base,spacing.c_str(),spacing.length());
	
	// Declare and store the trivia, then create the token to add.
	Trivia* t = _mem.emplace<Trivia>(TriviaKind::Whitespace,std::string_view((char*)spacing_base,spacing.length()));
	tok_list.push_back(tok.withTrivia(_mem,{t,1}));
}

// Update the modifiers token list
work->modifiers = TokenList(tok_list);

std::cout << "Modifier post-update: " << work->modifiers.toString() << std::endl;
```

I feel that this is quite convoluted for "not much" but didn't find an example of code that would do this (as slang code works on the buffers that are properly stored at once). Would you have some insight ? 

Anyway, on the line
```systemverilog
const logic [1:0] foo;
```
this produces
```
Modifier pre-update: 
const
Modifier post-update:  !-const
```
Therefore, I feel that I am on the right path !

Thanks a lot for all the help you're providing !

**Reply by MikePopoloski** [reply, marked as answer]
You have the right idea. It's going to be clunky because the memory management is important, but once you have it figured out you can wrap it all up in some nice helper methods and make it look pretty for your actual rewriter logic so I wouldn't worry about it too much.

Some thoughts: if the string you're adding is an actual string literal (in this case " !=") then you don't need to allocate memory for it; string literals live in a fixed location in memory for the lifetime of the program so you can just pass a pointer to it without issue. You only need the alloc and copy for dynamically constructed strings.

Also you need to be careful about this line:
```
work->modifiers = TokenList(tok_list);
```
That saves a pointer to a local vector on the stack, so if you leave this function it will become garbage. You need to make sure that memory lives dynamically, again by using the BumpAllocator. If you use slang's SmallVector for this there is a handy utility method `copy` that will do what you need (this is a common idiom throughout the library):
```
BumpAllocator alloc;
SmallVector<Token> buffer;
for (...) {
    buffer.push_back(...);
}
work->modifiers = buffer.copy(alloc);
```

**Reply by suzizecat** [reply]
> Some thoughts: if the string you're adding is an actual string literal (in this case " !=") then you don't need to allocate memory for it; string literals live in a fixed location in memory for the lifetime of the program so you can just pass a pointer to it without issue. You only need the alloc and copy for dynamically constructed strings.

Thanks for the heads up. I am not very familiar with `string_views` and a fair bit of the modern C++ stuff. In this situation, the spacing will be dynamically generated anyway.

> Also you need to be careful about this line:
> 
> ```
> work->modifiers = TokenList(tok_list);
> ```
> That saves a pointer to a local vector on the stack, so if you leave this function it will become garbage. You need to make sure that memory lives dynamically, again by using the BumpAllocator.

Yes, I saw it after posting. 

Thanks for the tip about the small vector ! I was wondering how to conveniently create a range and such.
So I'll wrap the clunky part into a nice function and it should be OK.

**Reply by suzizecat** [reply]
The final kind of result I have is functions such as this one: 
```cpp
Token replace_spacing(slang::BumpAllocator& alloc , const Token& tok, const unsigned int space_number)
{
	// Store the string in "persistent" memory (BumpAllocator)
	if(space_number == 0)
		return tok.withTrivia(alloc,{});
	
	slang::byte* spacing_base = alloc.allocate(space_number, alignof(char));
	std::memset(spacing_base,' ',space_number);

	// Declare and store the trivia, then create the token to add.
	Trivia* t = alloc.emplace<Trivia>(TriviaKind::Whitespace,std::string_view((char*)spacing_base,space_number));
	return tok.withTrivia(alloc,{t,1});
}
```
This function should be used with caution (as it may remove too much trivias) but will allow to set the spacing of a single token easily.

With that done, using the modifiable version of the AST, created through `clone`, I can simply replace the targeted Token while being sure to not remove any meaningful content.

This is made with something like:
```cpp
ast_node->token = replace_spacing(alloc, ast_node->token, number_of_spaces);
```

With that said, and with some more code, I get to convert:
```systemverilog
// Before:

logic single_unpacked [3:0];
logic [1:0][K_DIM-1:0] multi_packed_param_and_more;
logic [1:THE_FOO][3:0] multi_packed;

logic multi_unpacked_weird [3][K_DIMENSION];
logic [1:0][15:0] full_mix [3][31];

logic simple;
logic [THE_LONG_BAR:0] single_packed; //! A comment

// Another signal
logic [1:0][K_DIM-1:0] multi_packed_param; //! Commented !

// This one is not aligned
    // This is a silly comment
static const logic  multi_unpacked [3:0][31:0];

logic multi_unpacked_single [3][31]; 
```

To 
```systemverilog
//After
 logic                        single_unpacked             [3:0];
 logic [1:0      ][K_DIM-1:0] multi_packed_param_and_more      ;
 logic [1:THE_FOO][3      :0] multi_packed                     ;
 
 logic             multi_unpacked_weird [3][K_DIMENSION];
 logic [1:0][15:0] full_mix             [3][31         ];
 
 logic                  simple       ;
 logic [THE_LONG_BAR:0] single_packed; //! A comment
 
 // Another signal
 logic [1:0][K_DIM-1:0] multi_packed_param; //! Commented !
 
 // This one is not aligned
 // This is a silly comment
 static const logic multi_unpacked [3:0][31:0];
```

So I would call this a success !

**Comment by suzizecat** [comment]
Hi there !

I was going again on this whole formatter thing and came across a new issue that I hadn't stumbled upon previously. 

I am currently working on a formatter "pass" that would only perform the basic indentation of the code.
Let says that my initial code is: 

```systemverilog
module a ();

logic g,h;
logic [1:0] i;

    assign {g, h} = i;

  always_ff @( posedge i_clk or negedge i_rst_n ) begin : p_seq_
        if (~i_rst_n) begin
            i <= 2'b0;
              end else begin
                    i <=    ~i;
        end
    end

endmodule
```

And the end result would look like 
```systemverilog

```systemverilog
module a ();

    logic g,h;
    logic [1:0] i;

    assign {g, h} = i;

    always_ff @( posedge i_clk or negedge i_rst_n ) begin : p_seq_
        if (~i_rst_n) begin
            i <= 2'b0;
        end else begin
            i <=    ~i;
        end
    end
endmodule
```
With one indentation level being 4-spaces.
My current issue is when handling the `always_ff` part. I am using a `SyntaxRewriter` and if I simply visit the `always_ff` through a `ProceduralBlockSyntax` handler, I can rewrite the `i <= ...` statements as wanted but, obviously, I don't indent the `always_ff` itself. 

Now, If I rewrite the `always_ff` itself, I don't lose the formatting of its content (while this formatting is still done somewhere, but not on the `ProceduralBlockSyntax` that is present in the final CST). 

My handler function is something like this:
```cpp
void IndentationPhaseVisitor::handle(const slang::syntax::ProceduralBlockSyntax& node)
{
	{
		spdlog::info("Increase indent due to kind {}",toString(node.kind));
		IndentLock idt(*_idt);
		visitDefault(node);
	}
	_default_indent_handler(node);
}


const SyntaxNode* IndentationPhaseVisitor::_default_indent_handler(const slang::syntax::SyntaxNode& node)
{
	Token first_token = node.getFirstToken();
	if(! first_token.isOnSameLine())
	{
		SyntaxNode* stx = deepClone(node,_mem);
		// Replace the first token by a new one with the proper indentation
		*(stx->getFirstTokenPtr()) = _idt->indent(first_token);
		replace(node,*stx);
		return stx;
	}

	return &node;
}
```

I also tried to *first* process the node, then visit the modified version (the return value of `_default_indent_handler`) ... Or visit the original version (node) without success.
My guess is that somehow I don't visit the node that is kept after `replace` but I don't see how to do that. 

Would you have some pointers ? 

Best regards,
Julien

**Reply by MikePopoloski** [reply]
The way the rewriter works is that it saves up a list of changes and then applies them all at the end to do the transformation. It's sort of assumed that the new nodes you're adding are in the format you want. If not I guess you need to force another visit pass on that subtree by like, invoking the rewriter again on it.

**Reply by suzizecat** [reply]
Okay, that makes sense. 
That was what I was trying to do by `defaultVisit`'ing the generated node instead of the original one to try and update it on the fly.

By any chance, wouldn't it be possible to force the "immediate" commit of pending actions ?

**Reply by suzizecat** [reply]
So, I just took the full iterative approach (flag if a node that you should visit has also been replaced, and then skip it. End the full process once you  don't have this situation anymore).

It is probably not the most efficient overall, but it is reliable and avoid a whole lot of problems.
Thanks for the pointer !

Just a side note, it is not possible to recursively use the visitor, as `transform` requires a `SyntaxTree` (not a `SyntaxNode` ) hence it cannot be called on an arbitrary subtree of the CST.

**Reply by MikePopoloski** [reply]
Makes sense. The logic SyntaxRewriter is doing isn't that complicated, you could write your own thing that does what you want if you need to.

**Comment by sknaumov** [comment]
@suzizecat Hello. Do I correctly understand that this discussion is about your [slang-lsp-tools](https://github.com/suzizecat/slang-lsp-tools) project? If yes, could you add a readme of how to try to use this tool as a standalone formatter?

**Reply by suzizecat** [reply]
@sknaumov Yes it is.

You may try the formatter by following the build instructions for slang-lsp but using the `sv-formatter-exe` target. That means that you should have something like so:

```shell
$ cmake -DCMAKE_BUILD_TYPE=Release -B./build
$ cmake --build ./build --target sv-formatter-exe
$ ./build/sv-formatter --help
```

You will then have the `sv-formatter` executable in the `build/` folder and you should be able to use the `--help` argument to get more info about how to use it. You will also find the executable in the releases artifact (built for Ubuntu).

If you want to look at the sources, you will have everything under the `formatter/` folder of my repository, with `main_formatter.cpp` being the entry point.

Please note that this is work in progress (as it is quite a tedious work and I had to work on other parts of the language server) so the final quality will probably not be on par with others. 
You are welcome to provide feedbacks though (preferably on [slang-lsp-tools](https://github.com/suzizecat/slang-lsp-tools) to avoid thread pollution ).

**Comment by suzizecat** [comment]
Hi there ! 
It has been a long time, but I am facing an issue when going from Slang 10.0 to slang 11.0 related to code discussed here, so I though I might as well add it here.

It was mentioned in https://github.com/MikePopoloski/slang/discussions/828#discussioncomment-8233513 that the following idiom was valid to build token lists: 
```cpp
BumpAllocator alloc;
SmallVector<Token> buffer;
for (...) {
    buffer.push_back(...);
}
work->modifiers = buffer.copy(alloc);
```

Which I used quite a bit. 

However, this now fails (on Slang v11.0) with:
```
error: no viable overloaded '='
        work->modifiers = tok_list.copy(_mem);
        ~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~
.../include/slang/syntax/SyntaxNode.h:658:20: note: candidate function (the implicit copy assignment operator) not viable: no known conversion from 'std::span<Token>' to 'const TokenList' for 1st argument
class SLANG_EXPORT TokenList : public SyntaxListBase<TokenList, parsing::Token> {
                   ^~~~~~~~~
.../include/slang/syntax/SyntaxNode.h:658:20: note: candidate function (the implicit move assignment operator) not viable: no known conversion from 'std::span<Token>' to 'TokenList' for 1st argument
class SLANG_EXPORT TokenList : public SyntaxListBase<TokenList, parsing::Token> {
``` 
(Where `tok_list` is a `slang::SmallVector<slang::parsing::Token> tok_list;`)

This also fails on `SeparatedSyntaxList` (on which it used to work just fine too)...
As I can't seems to find what was the breaking change, would someone have a pointer ?

Best regards !

**Reply by suzizecat** [reply]
So, of course, I found it 5 minutes after asking the question...
My solution is to replace
```cpp
work->modifiers = buffer.copy(alloc);
```
By
```cpp
work->modifiers = {alloc, buffer};
```

Could you confirm that this is actually "the right way" to do it ? 
If so (and if this is not a bug), it might be worth mentioning as breaking changes in the release notes.

**Reply by MikePopoloski** [reply]
Yes, the syntax list constructors now take an allocator and copy the elements inline.

Since so much of the slang API is exposed to downstream users, pretty much every change is potentially breaking to somebody, so it's hard to capture all of that in the release notes without making them uselessly verbose.
